vividus-plugin-mail/src/main/java/org/vividus/bdd/mail/service/ImapMessageFetchService.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright 2019-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.bdd.mail.service;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import javax.mail.Authenticator;
+import javax.mail.FetchProfile;
+import javax.mail.FetchProfile.Item;
+import javax.mail.Folder;
+import javax.mail.Message;
+import javax.mail.MessagingException;
+import javax.mail.PasswordAuthentication;
+import javax.mail.Session;
+import javax.mail.Store;
+import javax.mail.event.MessageCountEvent;
+import javax.mail.event.MessageCountListener;
+import javax.mail.search.SearchTerm;
+
+import org.vividus.bdd.mail.factory.EmailMessageFactory;
+import org.vividus.bdd.mail.factory.EmailMessageFactory.EmailMessageCreationException;
+import org.vividus.bdd.mail.model.EmailMessage;
+import org.vividus.bdd.mail.model.MailServerConfiguration;
+import org.vividus.util.Sleeper;
+import org.vividus.util.wait.WaitMode;
+import org.vividus.util.wait.Waiter;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+public class ImapMessageFetchService implements IMessageFetchService
+{
+    private static final String PROTOCOL = "imaps";
+
+    private final WaitMode waitMode;
+    private final Duration messageEventWaitDuration;
```

Comment:
I believe it should be calculated automatically and be based on `waitMode`

Reply:
these values are having a bit different meaning, generally we have two points of wait
1) message is sent from external service to recipient
2) java client observes new message arrives to recipient
So the waitMode is responsible for sending sync signal for new messages, while messageEventWaitDuration is responsible for getting new messages due to that signal
I think this values should be lower than polling time and probably it makes sense to hardcode it

Reply:
decided to use coefficient